utftp.1
.TH utftp 1
.SH NAME
utftp \- a simple TFTP client
.SH SYNOPSIS
.B utftp
[\fIoptions\fR] SERVER[:PORT] remote-file-name [local-file-name]
.SH DESCRIPTION
.B utftp 
sends and receives files with the TFTP protocol.

The default is to receive files. To change that use the 
.I \-\-put 
option, or rename the program so that it's name contains
the three characters 
.I "put"
(or create a link: 
.IR "ln utftp utftp-put" ).

.SH OPTIONS
.TP
.B "\-b SIZE, \-\-blocksize SIZE" 
try to use blocks of SIZE size instead of 512. Size may be between 512
and 65464, but i don't believe something larger then 32768 will work
(aside from that it wouldn't be that useful). Depending on you system
you may not be able do get anything transfered with blocksizes as 
low a 4096.

Common UNIX tftp servers ignore all options. In that case 512 will be
used, regardless of your wishes.

.BR "Note" :
this client is not intended for real use now. It logs all errors through
syslogd, and it's interface is rather ugly. But you might find it useful
anyway: it knows about the blocksize option.

.TP
.B "\-g, \-\-get" 
get a file from the server.

.TP
.B "\-n, \-\-netascii" 
Use 
.I netascii
conversion (this doesn't hurt, except for performance, and is not really useful).

.TP
.B "\-p, \-\-put" 
put a file on the server.

.TP
.B "\-t TIMEOUT, \-\-timeout TIMEOUT" 
use a timeout (for retries) of TIMEOUT seconds instead of 5. TIMEOUT
must be between 1 and 255 inclusively.

utftp will send an 
.I timeout
option to the server (to make it use the same timeout value), but
common UNIX tftp servers ignore all options, in which case the server
will still use a timeout of 5 seconds (usually).

.PP
In case i screwed something up:
short options use the same arguments as the long ones

.SH EXITCODES
utftp will exit using one of the following exit codes:
.TP
.B "0"
OK, action completed as requested.
.TP
.B "1"
local error (out of memory, can't create/read file, and so on)
.TP
.B "2"
usage error.
.TP 
.B "3"
timed out.
.TP
.B "4"
communication problem (TFTP protocol error).
.TP
.B "5"
network problem.
.TP
.B "6"
received some kind of 
.I "error"
(like access denied).
.TP
.B "7"
undefined.

.SH "SEE ALSO"
RFC 1350, RFC 2349, RFC 2348.


NEWS
0.2.4 - 1 November 1999

* some cleanup and minor bug fixes
* utftpd left SCCS lock files laying around in case of a failed put
  request.

0.2.3 - 5 September 1999

* the server now ignores garbage bytes at the end of a packet, unless
  there were tftp options.
* "must not write as root" check moved some lines up, to check against
  this before we create files (old version left 0 length files).

0.2.2 - 22 August 1999

* the server now uses adaptive timeout.
* the server now unlinks the received file if it was not received 
  completely.
* the server now includes a workaround for the 
  "Sorcerer's Apprentice Syndrome" (as it's called by RFC 1123)

0.2.1 - 2 May 1999

* rewrote configuration file handling. It's now much more 
  beautiful. 
  See sample.config for more information.
  The old stuff partially continues to work, except that
* the \ at the end of line (to split long lines into smaller pieces)
  will not work anymore. Sorry for that, but it's just too much
  pain to fix this.

0.2.0 - 2 May 1999
* understands a new experimental tftp option (revision).
* major internal interface cleanup.
* now reads version controlled files from a pipe. This means that
  it now can serve read request even when the same file is just uploaded.
* "lying around" copies of version controlled files will now be ignored.
* -l option (GNU inetutils) is now recognized (and ignored,
  utftpd logs by default).
* -n option (GNU inetutils) implemented. It's used to suppress NAK
  answers for "file not found". 
* Can now operate without a configuration file, in a mode largely 
  compatible with the traditional UNIX tftp server: Access control
  based on a list of directories given on the command line and the
  UNIX access bits (read/write for others). Therefore
* it may now be used as a drop-in replacement for GNU inetutils tftpd.

0.1.5 - 10 April 1999
* added RCS support. See the utftpd.conf manual page for a description 
  of the new configuration variables rcs-ci and rcs-co.
* improved SCCS support: unchecked files will not be checked in 
  (that's useful if your router saves his configuration regulary even 
  if it's unchanged). See the utftpd.conf manual page for a 
  description of the new configuration file variables sccs-clean and
  sccs-unget.
* improved checksuite to check SCCS/RCS handling.
* fixed some small bugs.
* improved compatibility (SCO 3.2v4.2, BSDI 3.1, Solaris 2.6 and 2.5.1,
  OpenBSD 2.4/alpha, HPUX 9.00/m68k).

0.1.4 - 5 April 1999
* configuration (cdb) file may now reside outside the chrooted area.
* understands RFC 2348 "blksize" option.
* understands RFC 2349 "timeout" option (but doesn't know about "tsize" 
  now).
* test suite added.
* documentation fixes.
* server now knows about wildcards.
* threw out some bugs and unused library files.
* there's now a client program (utftp).

0.1.3 - 28 March 1999
* uh, 0.1.2 didn't work because of those stupid last-moment-large-rewrite ... 
  Ok, that's why i tested it on my companies tftp server.
* smaller log things.

0.1.2 - 28 March 1999
* manual pages
* rpm fixed

0.1.1 - 21 March 1999
* put on cvs


utftp.c
/* utftp.c: simple tftp client to put files */
/*
 * Copyright (C) 1999 Uwe Ohse
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#include "config.h"
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <signal.h>
#include <fcntl.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include "no_tftp.h"
#include <netdb.h>
#include <pwd.h>
#include <grp.h>

#include <setjmp.h>
#include <syslog.h>
#include <errno.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "uogetopt.h"
#include "timselsysdep.h"
#include "nonblock.h"
#include "str2num.h"
#include "uostr.h"
#include "uoio.h"
#include "cdb.h"
#include "sys/wait.h"
#include "wildmat.h"
#include "tftplib.h"


static const char *argv0;
static char *opt_revision;
static uoio_t io_e;
static int opt_netascii=0;
static int opt_is_put=0;
static int opt_stop=0;
static int opt_duplicate_ack=0;
static unsigned long opt_force_timeout=0;
static unsigned long opt_segsize=512;
static unsigned long opt_timeout=5;
static unsigned long opt_retries=5;
static int opt_send_garbage=0;

static uogetopt_t myopts[]={
	{'b',"blocksize",   UOGO_ULONG,&opt_segsize,0, "use a blocksize different from the default 512.\n",0},
	{'d',"duplicate-ack",   UOGO_FLAG,&opt_duplicate_ack,1, 
		"duplicate ACKs sent\n",0},
	{'f',"force-timeout",   UOGO_ULONG,&opt_force_timeout,0, 
		"test server timeout handling\n"
		"by sleeping some seconds","SECONDS"},
	{'G',"garbage",   UOGO_FLAG,&opt_send_garbage,1, 
		"send 2 garbage bytes after the request (test only)\n",0},
	{'g',"get",   UOGO_FLAG,&opt_is_put,0, "get file from server (default)\n",0},
	{'n',"netascii",   UOGO_FLAG,&opt_netascii,1, "use netascii conversion\n",0},
	{'p',"put",   UOGO_FLAG,&opt_is_put,1, "put file onto server\n",0},
	{'r',"revision",  UOGO_STRING,&opt_revision,1, "get or put revision REV\n","REV"},
	{'R',"retries",   UOGO_ULONG,&opt_retries,0, "number of retries on timeout (1..255, default 5).\n",0},
	{'S',"stop",   UOGO_FLAG,&opt_stop,1, "stop upload after the first block.\n",0},
	{'t',"timeout",   UOGO_ULONG,&opt_timeout,0, "timeout value (1..255, default 5).\n",0},
    {0,0,0,0,0,0,0}
};

static void
die (int e, const char *where)
{
	uoio_write_cstr (&io_e, argv0);
	uoio_write_cstr (&io_e, ": ");
	uoio_write_cstr (&io_e, where);
	if (e) {
		uoio_write_cstr (&io_e, ": ");
		uoio_write_cstr (&io_e, strerror (e));
	}
	uoio_write_cstr (&io_e, "\n");
	uoio_flush (&io_e);
	_exit (1);
}


static void 
usage(void)
{ die(0,"Give --help for usage information"); _exit(1); }

int 
main (int argc, char **argv)
{
	int port;
	struct servent *sp;
	char *p;
	struct tftplib_ctrl *ctrl;
	const char *target;
	const char *source;
#ifdef HAVE_LIBEFENCE
    /* hack around efence. Don't care about error messages here */
	{int fd,fd2;void *waste;
	 if (-1==(fd=open("/dev/null",O_WRONLY))) _exit(1);
	 if (-1==(fd2=dup(2))) _exit(1);
	 if (-1==(dup2(fd,2))) _exit(1);
	 waste=malloc(1);
	 if (-1==(dup2(fd2,2))) _exit(1);
	 close(fd); close(fd2);
	}
#endif

	argv0 = strrchr (argv[0], '/');
	if (argv0)
		argv0++;
	else
		argv0 = argv[0];
	uoio_assign_w (&io_e, 2, write, 0);
	if (strstr(argv0,"put")) opt_is_put=1;

	uogetopt ("utftp", PACKAGE, VERSION, &argc, argv, uogetopt_out,
			  "usage: utftp host[:port] filename-on-host [local-file-name]",
			  myopts,
			  "  \"local-file-name\" defaults to \"filename-on-host\" without any directory part.\n"
			  "  Report bugs to uwe-utftpd@bulkmail.ohse.de");

	if (!argv[1] || !argv[2] || (argv[3] && argv[4]))
		usage ();
	if (opt_segsize<512) {
		/* this can't work: if the server doesn't know about the blksize option it will send 512 byte blocks */
		die(0,"blocksize to low. 512 is minimum");
	}
#ifdef LOG_PERROR
	openlog(argv0,LOG_PERROR|LOG_PID,LOG_USER);
#else
	/* damned, HPUX 9.00 doesn't have LOG_PERROR. */
	openlog(argv0,LOG_PID,LOG_USER);
#endif
	if (opt_timeout<1 || opt_timeout >255) die(0,"illegal timeout value");
	if (opt_retries<1 || opt_retries >255) die(0,"illegal retry value");

	p=strchr(argv[1],':'); 
	if (p) {
		unsigned long ul;
		*p++=0;
		if (-1==str2ulong(p,&ul,0)) die(0,"not a valid port number");
		port=ul;
		if (ul!=(unsigned long) port) die(0,"not a valid port number");
		port=htons((unsigned short)port);
	} else {
		sp = getservbyname ("tftp", "udp");
		if (sp == 0) die (0, "udp/tftp: unknown service");
		port=sp->s_port;
	}
	
	ctrl=tftp_setup_ctrl(0, opt_segsize);
	if (!ctrl) die(0,"out of memory");
	ctrl->netascii=opt_netascii;
	ctrl->timeout=opt_timeout;
	ctrl->retries=opt_retries;
	ctrl->send_garbage=opt_send_garbage;
	if (opt_revision) ctrl->revision=opt_revision;
	source=argv[2];

	target=argv[3];
	if (!target || 0==strcmp(target,"--")) {
		target=strrchr(source,'/'); 
		if (target) target++;
		else target=source;
	}
	ctrl->force_timeout=opt_force_timeout;
	ctrl->force_stop=opt_stop;
	ctrl->duplicate_ack=opt_duplicate_ack;

	if (opt_is_put)
		return tftp_send(argv[1],port,source,target,ctrl);
	else
		return tftp_receive(argv[1],port,source,target,ctrl);
}
